metsim/metsim.py

```diff
@@ -48,129 +79,152 @@ def __init__(self, params: dict):
+        out_dir = MetSim.params['out_dir']
+        if not os.path.exists(out_dir):
+            os.makedirs(out_dir)
+
+        self.output = self.domain.copy(deep=True)
+        self.output.attrs = attrs['_global']
+        # Number of timesteps
+        if self.disagg:
+            delta = pd.Timedelta('1 days')
+        else:
+            delta = pd.Timedelta('0 days')
+        times = pd.date_range(MetSim.params['start'],
+                              MetSim.params['stop'] + delta,
+                              freq="{}T".format(MetSim.params['time_step']))
+        n_ts = len(times)
+
+        shape = (n_ts, ) + self.domain_shape
+        dims = ('time', ) + self.domain_dims
+        coords = {'time': times, **self.domain.mask.coords}
+        for varname in MetSim.params['out_vars']:
+            self.output[varname] = xr.DataArray(
+                data=np.full(shape, np.nan),
+                coords=coords, dims=dims,
+                name=varname, attrs=attrs.get(varname, {}),
+                encoding=None)
+        # Input preprocessing
+        in_preprocess = {"ascii" : self.vic_in_preprocess,
+                         "binary" : self.vic_in_preprocess,
+                         "netcdf" : self.netcdf_in_preprocess}
+        in_preprocess[MetSim.params['in_format']](job_list)
+
+        # Output preprocessing
+        out_preprocess = {"ascii": self.ascii_out_preprocess,
+                          "netcdf": self.netcdf_out_preprocess}
+        out_preprocess[MetSim.params['out_format']]()
+        self.ready = True
 
     def launch(self, job_list):
         """Farm out the jobs to separate processes"""
-        # Used to dispatch to the correct preprocessing functions
-        in_preprocess = {"ascii": self.ascii_in_preprocess,
-                         "binary": self.binary_in_preprocess,
-                         "netcdf": self.netdf_in_preprocess}
-        met_data = in_preprocess[MetSim.params['in_format']](job_list)
-
-        # Start up the IO Process
-        io_process = Process(target=self.create_io_process, args=(self.queue,))
-        io_process.start()
-
-        # Split the jobs up into groups based on the number
-        # of desired processes
-        self.locations = np.array_split(np.array(self.locations),
-                                        MetSim.params['nprocs'])
-        process_handles = [Process(target=self.run, args=(met_data, locs,
-                                   self.queue)) for locs in self.locations]
-
-        # Runs everything
-        for p in process_handles:
-            p.start()
-        for p in process_handles:
-            p.join()
-
-        # Close everything out
-        self.queue.put("done")
-        io_process.join()
-
-    def run(self, met_data, locations, queue):
+        nprocs = MetSim.params['nprocs']
+        self.pool = Pool(processes=nprocs)
+
+        # TODO: fix the hard coded 10
+        locations = np.array_split(list(zip(self.i_idx, self.j_idx)), nprocs * 10)
```

Comment:
Can we implement a simple heuristic for determining chunk size here? I'm not sure exactly what that would be...